### ISSUE TITLE
Provide input repo names as env variables

### ISSUE THREAD

**Issue by brollb** [pachyderm/pachyderm#1785]
#### tl;dr:
When I am working on a pipeline spec, I sometimes find myself updating the input repo and re-running it only to have it fail because I forgot to update the input repo name in the bash command used to run the pipeline (usually when listing files with something like `find /pfs/input_repo_name`).

It might be nice to have the inputs exposed as env variables (analogously to `$1`, `$2`, etc in bash scripts) like `$INPUT_REPO_1`, `$INPUT_REPO_2` or something.

#### Here a more concrete example:
I have some command `my-command` which takes filenames and performs some sort of operation on them and saves them to a specified output dir (say `-o /pfs/out`). In pachyderm pipelines, I script this as follows (assuming the input repo is `input_repo`):

```
my-command $(find /pfs/input_repo -not -path /pfs/input_repo) -o /pfs/out
```

In this command, the nested command will return the absolute paths for each of the files in the repo (excluding the main directory) which will then be processed by `my-command`. My pipeline spec then might look something like this:

```json
{
  "pipeline": {
    "name": "example-pipeline"
  },
  "transform": {
    "image": "ubuntu",
    "cmd": ["sh"],
    "stdin": [
      "my-command $(find /pfs/input_repo -not -path /pfs/input_repo) -o /pfs/out"
    ]
  },
  "inputs": [
    {
      "repo": {
        "name": "input_repo"
      },
      "glob": "/*"
    }
  ]
}
```

Anyway, now I sometimes find myself renaming the input repo (or wanting to run the same pipeline on a different input). Naturally, I change the repo name in the `inputs` field (and the pipeline name) and re-run. However, as I forgot to change the occurrences of `input_repo` in the `stdin` field of the transform, the pipeline fails and I need to reupdate that field and re-run.

It might be nice to be able to reference the inputs using env variables so the actual inputs passed to the pipelines won't break the transform command. In this case, the pipeline might look something like:

```json
{
  "pipeline": {
    "name": "example-pipeline"
  },
  "transform": {
    "image": "ubuntu",
    "cmd": ["sh"],
    "stdin": [
      "my-command $(find /pfs/$INPUT_REPO_1 -not -path /pfs/$INPUT_REPO_1) -o /pfs/out"
    ]
  },
  "inputs": [
    {
      "repo": {
        "name": "input_repo"
      },
      "glob": "/*"
    }
  ]
}
```

This could also work for multiple inputs as they would be referenced using `$INPUT_REPO_2`, `$INPUT_REPO_3`, etc

**Comment by brollb** [comment]
I think this would be especially relevant once we have created an effective pipeline and want to run in on a new dataset (input repo). Currently, you would have to change the repo name and potentially change it in the stdin as well.

It might be nice to go one step further and be able to create a new pipeline from an existing spec and just changing the input repos:

```
pachctl create-pipeline -f original_pipeline.json input_repo1 input_repo2
```

Or perhaps more generally just overriding fields in the pipeline spec
```
pachctl create-pipeline -f original_pipeline.json --set inputs[0].repo.name=input_repo_1 --set pipeline.name=myNewPipeline
```

**Comment by derekchiang** [comment]
+1.  In fact, once we support the `$INPUT_REPO` environment variable, we don't even need the `/pfs` prefix anymore.  That way, internally we can mount the input data wherever we want, which also means that we can process multiple datums concurrently on the same worker, since there won't be any path collision.

**Comment by jdoliner** [comment]
After talking internally a bit I think there's already a pretty good way to solve this problem with the existing system by giving an explicit name to your inputs. Using this feature your spec would look like this:

```json
{
  "pipeline": {
    "name": "example-pipeline"
  },
  "transform": {
    "image": "ubuntu",
    "cmd": ["sh"],
    "stdin": [
      "my-command $(find /pfs/data -not -path /pfs/data) -o /pfs/out"
    ]
  },
  "inputs": [
    {
      "repo": {
        "name": "input_repo"
      },
      "glob": "/*",
      "name": "data"
    }
  ]
}
```

With this spec you could freely edit the name of the repo you're using and the code would keep working.

I prefer this to having env vars for each input because it's a little bit more explicit since the names are meaningful to humans and can be used to document what this input data means to the transform code. Furthermore I think you'll get a more subtle version of this same issue with env vars when you add a new input. Adding a new input means that some of your inputs will go from being `INPUT_N` to `INPUT_N+1` and you'll have to figure out which inputs should be incremented and which shouldn't. With explicit names you avoid this problem.

Does this work for your needs @brollb?

**Comment by brollb** [comment]
Yeah, I think that should work - thanks!
